controllers/topology/suite_test.go

```diff
@@ -17,21 +17,74 @@ limitations under the License.
 package topology
 
 import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
-	ctrl "sigs.k8s.io/controller-runtime"
-
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/api/v1beta1/index"
+	tlog "sigs.k8s.io/cluster-api/controllers/topology/internal/log"
+	"sigs.k8s.io/cluster-api/internal/envtest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
 )
 
 var (
+	timeout    = time.Second * 30
 	ctx        = ctrl.SetupSignalHandler()
 	fakeScheme = runtime.NewScheme()
+	env        *envtest.Environment
+	log        tlog.Logger
 )
 
 func init() {
 	_ = clientgoscheme.AddToScheme(fakeScheme)
 	_ = clusterv1.AddToScheme(fakeScheme)
 	_ = apiextensionsv1.AddToScheme(fakeScheme)
 }
+func TestMain(m *testing.M) {
+	setupIndexes := func(ctx context.Context, mgr ctrl.Manager) {
+		if err := index.AddDefaultIndexes(ctx, mgr); err != nil {
+			panic(fmt.Sprintf("unable to setup index: %v", err))
+		}
+	}
+	log = tlog.LoggerFrom(ctx)
+	setupReconcilers := func(ctx context.Context, mgr ctrl.Manager) {
+		unstructuredCachingClient, err := client.NewDelegatingClient(
+			client.NewDelegatingClientInput{
+				// Use the default client for write operations.
+				Client: mgr.GetClient(),
+				// For read operations, use the same cache used by all the controllers but ensure
+				// unstructured objects will be also cached (this does not happen with the default client).
+				CacheReader:       mgr.GetCache(),
+				CacheUnstructured: true,
+			},
+		)
+		if err != nil {
+			os.Exit(1)
```

Comment:
Can we panic with an error message instead (like in controllers/)? This way we have an error message to debug.

(same in l.78)